Utilisation de Git avec Rstudio (#219)
InseeFrLab/utilitR: 03_Fiches_thematiques/Fiche_configurer_git.Rmd

# Configurer `Git` sur son poste de travail {#git-config}

## Tâches concernées et recommandations

> L'utilisateur souhaite versionner son projet depuis son poste de travail en ayant recours pour cela à un dépôt distant, disponible sur un espace type `Gitlab` ou `Github`. 

Reviewer: utiliser les graphies officielles

```suggestion
L'utilisateur souhaite versionner son projet depuis son poste de travail en ayant recours pour cela à un dépôt distant, disponible sur un espace type `GitLab` ou `GitHub`. 
```

L'utilisation du logiciel `Git` en local, qui relève plutôt de l'usage quotidien du logiciel, est reportée dans une fiche dédiée ([Utiliser Git avec RStudio](#git)). 

::: recommandation
* Pour interagir avec un dépôt `Git` distant, l'utilisateur a le choix

> entre deux protocoles d'authentification: `SSH` ou `HTTPS`.

Reviewer: il me semblait qu'on avait dit qu'on utilisait désormais des espaces avant `:`

```suggestion
entre deux protocoles d'authentification : `SSH` ou `HTTPS`.
```

* Il est conseillé d'utiliser le protocole  `HTTPS` qui est simple à configurer. 
Vous pouvez utiliser le protocole `SSH` si vous le souhaitez, mais il peut s'avérer difficile à configurer.
* Si vous utilisez  l'authentification `HTTPS`, il est recommandé 
de s'authentifier avec des jetons personnels d'accès (*personnel access tokens*) plutôt qu'avec 
un _login_ et un mot de passe.
:::

## Les différentes forges utilisables

> Une `r with_def("forge")` est un espace informatique en ligne qui accueille des dépôts distants. Un dépôt distant (_remote repository_ en anglais) est un dossier que vous pouvez utiliser pour stocker vos codes (et leur historique), et pour partager des codes avec vos collègues ou stocker vos codes personnels.

Reviewer: Attention :  `r with_def("forge")` donne **forges** (au pluriel) dans la preview

Une forge prend généralement forme d'un site internet. Il existe deux forges principales :

> * Github (accessible sur le Web, à l'adresse <https://github.com>), qui appartient à une entreprise privée ;

Reviewer: utiliser la graphie officielle et tant qu'à faire, on peut préciser que c'est MS

```suggestion
* GitHub (accessible sur le Web, à l'adresse <https://github.com>), qui appartient à Microsoft ;
```

> * Gitlab qui se décline sous différentes formes :

Reviewer: utiliser la graphie officielle

```suggestion
* GitLab qui se décline sous différentes formes :
```

>     + une version publique, disponible sur <https://gitlab.com>, qui propose les fonctionnalités principales de versionnage du code mais également des fonctionnalités supplémentaires (discussions, mise à disposition de machines (*runners*) pour tester des scripts...) ;

Reviewer: Pourquoi le préciser uniquement pour GitLab et pas pour GitHub ? on croirait à la lecture que ce n'est pas le cas pour GitHub

Author: ça c'est un ajout d'@oliviermeslin. Oui c'est *confusing*

Author: je pense que l'ajout de "qui appartient à une entreprise privée" n'est pas nécessaire. Sinon il faut l'ajouter à gitlab et expliquer alors que le gitlab.com c'est privé mais avec des fonctionalités opensource et d'autres qui ne le sont pas

Reviewer 2: OK

    + des versions internes, qui reprennent la majorité des fonctionnalités de la version publique mais proposent des machines (appelées *runners*) internes. Celles-ci ne sont pas forcément ouvertes sur internet.   

Les forges proposent des fonctionnalités précieuses pour gérer un projet impliquant du code ou de la documentation. La formation
[Travail collaboratif avec R](https://linogaliana.gitlab.io/collaboratif/git.html) décrit ces fonctionnalités de manière plus détaillée.

::: specificite

> Une forge interne `Gitlab` est disponible sous AUS. Son adresse, ainsi que des éléments complémentaires à cette fiche, sont disponibles dans la documentation AUS (`Y:/Documentation/AUSV3/`).

Reviewer: précisions

```suggestion
Une forge interne `GitLab` est accessible depuis AUS. Son adresse, ainsi que des éléments complémentaires à cette fiche, sont disponibles dans la documentation AUS (`Y:/Documentation/AUSV3/`).
```

:::

::: remarque

> Une instance `Gitlab` est disponible sur le SSP Cloud [https://git.lab.sspcloud.fr/] avec laquelle peut interagir un service `RStudio` [lien vers fiche quand elle sera là], et vous pouvez l'utiliser si vous le souhaitez. Toutefois, le SSP Cloud est une plateforme connetée à Internet et a donc accès aux dépôts hébergés sur <https://gitlab.com> ou <https://github.com>. Il est donc conseillé d'utiliser ces deux `r with_def("forges")` plutôt que <https://git.lab.sspcloud.fr/> dont la pérennité n'est pas assurée à long terme.

Reviewer: ```suggestion
Une instance `GitLab` est disponible sur le SSP Cloud [https://git.lab.sspcloud.fr/] avec laquelle peut interagir un service `RStudio` [lien vers fiche quand elle sera là], et vous pouvez l'utiliser si vous le souhaitez. Toutefois, le SSP Cloud est une plateforme accessible depuis Internet et a donc accès aux dépôts hébergés sur <https://gitlab.com> ou <https://github.com>. Il est donc conseillé d'utiliser une de ces deux `r with_def("forges")` plutôt que <https://git.lab.sspcloud.fr/>.
```

:::

La suite de cette fiche décrit l'utilisation d'un dépôt distant sur une forge de type

> `Gitlab`. Néanmoins, la procédure pour configurer l'interaction avec un

Reviewer: ```suggestion
`GitLab`. Néanmoins, la procédure pour configurer l'interaction avec un
```

> dépôt distant situé sur `Github` est très similaire, ce qui est illustré

Reviewer: ```suggestion
dépôt distant situé sur `GitHub` est très similaire, ce qui est illustré
```

dans la fiche [Utiliser Git avec RStudio](#git).

## Installer Git sur son poste de travail {#git-install}

> Pour interagir avec un dépôt distant sur `Gitlab` ou `Github`, il est nécessaire

Reviewer: ```suggestion
Pour interagir avec un dépôt distant sur `GitLab` ou `GitHub`, il est nécessaire
```

> d'utiliser un outil de contrôle de version. Le plus utilisé est le logiciel `Git`.

Reviewer: ```suggestion
d'utiliser l'outil de contrôle de version `Git`.
```

> **Avant d'essayer d'installer `Git`, le premier réflexe est de vérifier qu'il n'est pas déjà disponible.** En effet, sur un serveur partagé comme AUS, ou sur une architecture comme le SSP Cloud, il n'est pas nécessaire d'installer `Git` puisque celui-ci est déjà disponible et configuré.

Reviewer: Est-il bien configuré sous AUS ?

Reviewer 2: Je vais voir avec Aurélien, mais je crois que oui

Si `Git` n'est pas disponible, il faut l'installer. Pour ce faire, il suffit
de télécharger le logiciel [site web officiel](http://git-scm.com/download/win), puis de l'installer. 
Une fois le logiciel installé, il est possible de le retrouver dans le menu Démarrer, rubrique Git.
L'utilisateur y trouvera plusieurs outils, dont un outil de ligne
de commande (`Git Bash`) et une interface (`Git GUI`). 

> ## Interaction avec un dépôt distant : principe

Reviewer: Pour que le titre apparaisse, faut rajouter un saut de ligne

```suggestion

## Interaction avec un dépôt distant : principe
```

[METTRE UN JOLI SCHEMA A PARTIR DE COLLABORATIF QUAND IL SERA PRET]

`Git` est un système décentralisé de contrôle de version : les codes sont
modifiés par chaque agent sur son poste de travail, puis sont mis en 
conformité avec la version collective disponible sur le dépôt distant au moment où l'agent le décide. 

Il est donc nécessaire que la forge connaisse l'identité de chacun des contributeurs, afin de déterminer qui est l'auteur d'une modification apportée aux codes stockés dans le dépôt distant. Pour que `Gitlab` reconnaisse un utilisateur suggérant des modifications, il est nécessaire de s'authentifier 
(un dépôt distant, même public, ne peut pas être modifié par n'importe qui). 
L'authentification consiste ainsi à fournir un élément que seul vous et 
la forge sont censés connaître : un mot de passe, une clé compliquée, un jeton d'accès...

> Plus précisément, il existe deux modalités pour faire connaître son identité à `Gitlab` :

Reviewer: ```suggestion
Plus précisément, il existe deux modalités pour faire connaître son identité à `GitLab` :
```

> * une **authentification HTTPS** ([décrite ici](git-connexion-https)) : l'authentification se fait grâce au login et au mot de passe, qu'il faut renseigner à chaque interaction avec le dépôt.

Reviewer 2: ```suggestion
* une **authentification HTTPS** ([décrite ici](git-connexion-https)) : l'authentification se fait avec un _login_ et un mot de passe (qu'il faut renseigner à chaque interaction avec le dépôt), ou avec un _token_ (méthode à privilégier).
```

* **une authentification SSH** : l'authentification se fait par une clé cryptée disponible sur le poste de travail et que `Github` ou `Gitlab` connaît. Une fois configurée, cette méthode ne nécessite plus de faire connaître son identité : l'empreinte digitale que constitue la clé suffit à reconnaître un utilisateur. 

::: conseil
Si vous utilisez l'authentification HTTPS, il est conseillé d'utiliser un jeton d'accès. Celui-ci permet de s'authentifier, est facilement révocable et ne bénéficie pas des super-pouvoirs qu'octroie un mot de passe comme changer le nom d'un dépôt voire le supprimer. Plus d'éléments sur les jetons d'accès sont disponibles [ici](https://happygitwithr.com/credential-caching.html#get-a-pat).
:::

::: remarque

> Ne jamais stocker un _token_, et encore moins son mot de passe, dans un projet. Un _token_ peut être stocké dans un système de gestion de mot de passe comme [Keepass](https://keepass.fr/).

Reviewer: il vaudrait mieux parler de credential helper : sous windows `git config --global credential.helper manager-core`

:::

> La méthode SSH peut être laborieuse à mettre en place, mais elle est plus confortable dans le travail quotidien car elle permet de ne pas avoir à renseigner son login et son mot de passe à chaque interaction avec le dépôt distant. Cependant, la méthode HTTPS est préférable car :

Reviewer: ce n'est pas exact si on utilise le credential helper

> * elle n'implique pas de configuration ;

Author: right, je modifie

```suggestion
* elle est beaucoup plus légère à configurer. L'utilisation d'un système de mémoire du mot de passe, à savoir le `credentiel helper`, est beaucoup plus simple à mettre en place que la génération de clés SSH ;
```

* elle n'est pas bloquée par des pare-feux, contrairement à la méthode SSH. Par exemple, il est impossible sur un poste Insee d'interagir avec un dépôt sur <https://gitlab.com> ou <https://github.com> en SSH ;

> * il est possible de conserver en mémoire ses identifiants un certain temps (grâce au `git credentials helper`) pour ne pas avoir à les renseigner à chaque fois.

Reviewer: ```suggestion
* il est possible de conserver en mémoire ses identifiants (grâce au `git credentials helper`) pour ne pas avoir à les renseigner à chaque fois.
```

## Configurer l'accès à dépôt distant en `HTTPS` {#git-connexion-https}

::: specificite

> Le `Gitlab` interne de l'Insee ne permet pas ce mode d'authentification. 

Reviewer: Ben si, avec un token

```suggestion
```

Author: J'ai modifié entre temps

> Il convient d'utiliser l'[authentification SSH](#git-connexion-ssh)

Reviewer: Pas d'accord, https+token

```suggestion
```

Author: Idem, j'ai modifié entre temps

:::

### Démarche générale

> Pour interagir avec un dépôt distant en utilisant le protocole `HTTPS`, il suffit de configurer le mot de passe dans `Gitlab`, en allant dans `Preferences > Password`, comme indiqué dans la capture d'écran ci-dessous :

Reviewer: ```suggestion
Pour interagir avec un dépôt distant en utilisant le protocole `HTTPS`, il suffit de configurer le mot de passe dans `GitLab`, en allant dans `Preferences > Password`, comme indiqué dans la capture d'écran ci-dessous :
```

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/password_conf.png", compression = FALSE)
```

Une fois le mot de passe configuré, il ne reste plus
qu'à `r with_def("cloner")` le dépôt distant sur lequel on veut travailler,
par exemple depuis `RStudio`, suivant le protocole décrit dans la fiche [Utiliser Git avec RStudio](#git), en ayant pris soin de sélectionner l'URL correspondant
au protocole `HTTPS`, comme ci-dessous :

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/clone_https.png", compression = FALSE)
```

Au moment du clônage, l'utilisateur est invité à entrer son identifiant
(celui de son profil Gitlab avec lequel il accède au dépôt distant), ainsi que

> son mot de passe (celui qu'il aura donc spécifié dans son compte Gitlab).

Reviewer: ```suggestion
son mot de passe (celui qu'il aura donc spécifié dans son compte Gitlab) auquel on substituera un token.
```

utilitr::include_image("./pics/config_git/config_pat_gitlab2.png", compression = FALSE)

Choisir un nom (par exemple `PAT_GITLAB`, mais ce n'est pas un nom obligatoire) et
des droits associés. Les droits `read_repository` et `write_repository`
permettent d'effectuer les opérations standard (_pull_ et _push_) décrites dans la fiche 
[Utilisation Git avec RStudio](#git).

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/config_pat_gitlab3.png", compression = FALSE)
```

Après avoir validé, le jeton s'affiche afin de la récupérer et le stocker dans
un endroit sécurisé (par exemple un gestionnaire de mot de passes). Attention,
ce sera la seule fois qu'il sera visible.

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/config_pat_gitlab4.png", compression = FALSE)
```

Plus bas sur la page, il est possible de voir l'ensemble des jetons utilisés, 
la date de dernière utilisation. C'est aussi ici que peut être supprimé le 
jeton en cas de doute sur sa confidentialité. 

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/config_pat_gitlab5.png", compression = FALSE)
```

::: remarque

> La marche à suivre pour créer un jeton d'accès sur `Github` est identique. La

Reviewer: ```suggestion
La marche à suivre pour créer un jeton d'accès sur `GitHub` est identique. La
```

[documentation officielle](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) présente des captures d'écran qui illustrent la démarche. Il faut choisir le 
niveau `repo` pour avoir les droits mentionnés précédemment. 
:::

### Garder en mémoire ses identifiants

L'authentification `HTTPS` implique en principe de renseigner ses identifiants (_login_ et
_password_, ou _token_) à chaque interaction avec le dépôt distant. Toutefois, `Git` propose
un outil pour conserver temporairement en mémoire (*cache*) des informations
d'authentification. **Cette mise en mémoire est complètement optionnelle, et vous pouvez utiliser  `Git` quotidiennement sans vous servir de cette possibilité.**

La mise en mémoire se fait en deux temps:

- Il faut d'abord ouvrir une invite de commande `Git Bash`
([explications ici](#terminal-git)) ;
- Il faut ensuite exécuter la ligne de code suivante pour stocker pendant quinze minutes les informations de connexion :

    ~~~~shell
    # Sous windows

>     git config --global credential.helper wincred

Reviewer: Comme discuté, je recommanderais plutôt `manager-core`

```suggestion
    git config --global credential.helper manager-core
```

    
    # Sous mac et linux

>     git config --global credential.helper 

Reviewer: ah bon ? aucun ? même pas `libsecret` sous linux ?

    ~~~~

::: remarque
Si vous avez fait une faute de frappe dans le mot de passe ou dans le jeton, il est possible de vider la mémoire
de la manière suivante :

~~~~shell

> git config --global --unset credential.helper

Reviewer: Si jamais vous retenez `manager-core`, je ne sais pas ce que donne cette commande (pas essayé)

~~~~
:::

## Configurer l'accès à dépôt distant en `SSH` {#git-connexion-ssh}

Pour interagir avec un dépôt distant en utilisant le protocole `SSH`,
il faut créer une clé SSH, plus exactement une paire de chaînes de caractères,
qui va permettre une authentification automatique auprès du dépôt distant.
La version publique de la paire de clés est connue du dépôt distant ; la
version privée reste quant à elle la propriété seule et unique de
l'utilisateur. C'est l'association par un logiciel de cryptographie
de ces deux versions qui permet l'authentification de l'utilisateur.
::: specificite

> L'authentification SSH ne fonctionne pas dans le Gitlab du SSP Cloud. Seule l'authentification en `HTTPS` est possible.

Reviewer: ```suggestion
L'authentification SSH ne fonctionne pas sur l'instance `GitLab` du SSP Cloud. Seule l'authentification en `HTTPS` est possible.
```

Un utilisateur peut générer autant de clés SSH qu'il le souhaite. Par exemple, vous pouvez générer une clé SSH par projet sur lequel vous travaillez (même si l'utilité de cette méthode n'est pas évidente).

Voici comment créer une clé SSH avec `Git Bash` :

- Il faut d'abord ouvrir une invite de commande `Git Bash`
([explications ici](#terminal-git)) ;
- Il faut ensuite exécuter la ligne de code suivante :
```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/ssh_keygen.png", compression = FALSE)
```

La clé est générée sur le chemin spécifié et sera constituée de deux fichiers texte :

- un fichier dont le nom se termine par `.pub` (comme _Public Key_) qui fournit la partie publique de la clé, et qui a vocation à être diffusée (donc c'est ce fichier qu'on chargera sur Gitlab pour permettre l'authentification) ;
- un fichier dont le nom se termine par `.ppk` (comme _Private Key_) qui est la partie privée de la clé, et qui ne doit **être diffusée en aucun cas**.

La version privée de la clé doit être localisée dans un dossier caché que Git va automatiquement utiliser (en général un dossier `.ssh` localisé sur un dossier personnel).

> ### Ajout de la clé SSH sous Gitlab

Reviewer: ```suggestion
### Ajout de la clé SSH sous GitLab
```

> Pour copier la partie publique de la clé sous Gitlab, il faut aller dans

Reviewer: ```suggestion
Pour copier la partie publique de la clé sous GitLab, il faut aller dans
```

> la partie `Settings` de l'espace personnel sous Gitlab, puis sélectionner

Reviewer: ```suggestion
la partie `Settings` de l'espace personnel sous GitLab, puis sélectionner
```

*"SSH Keys"* sur le bandeau latéral gauche. On se trouve face à une fenêtre
de texte comme ceci :

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/ssh_key_record.png", compression = FALSE)
```

dans laquelle on doit copier-coller le texte que l'on trouve dans le
fichier `.pub` de la clé SSH (en prenant soin d'éliminer l'éventuel texte
en fin du type *"imported-openssh-key"*).

> Gitlab est désormais configuré, et il est alors possible de faire

Reviewer: ```suggestion
GitLab est désormais configuré, et il est alors possible de faire
```

> dialoguer Git et Gitlab en suivant, par exemple, les routines proposées dans

Reviewer: ```suggestion
dialoguer Git et GitLab en suivant, par exemple, les routines proposées dans
```

# Configurer `Git` sur son poste de travail {#git-config}

> ## Tâches concernées et recommandations
> 
> L'utilisateur souhaite versionner son projet depuis son poste de travail en ayant recours pour cela à un dépôt distant, disponible sur un espace type `Gitlab` ou `Github`. 
> 
> L'utilisation du logiciel `Git` en local, qui relève plutôt de l'usage quotidien du logiciel, est reportée dans une fiche dédiée ([Utiliser Git avec RStudio](#git)). 
> 
> ::: recommandation
> * Pour interagir avec un dépôt `Git` distant, l'utilisateur a le choix
> entre deux protocoles de connexion, `SSH` ou `HTTPS`: 
> * Le
> protocole `SSH` est déconseillé en faveur du protocole `HTTPS` (qui ne requiert
> que de configurer un mot de passe sur le dépôt distant) ;
> * Plutôt qu'utiliser un mot de passe pour s'authentifier en `HTTPS`, il est
> recommandé de s'authentifier avec des jetons personnels d'accès
> (*personnel access tokens*).
> :::

Reviewer 2: ```suggestion
## Tâches concernées et recommandations

L'utilisateur souhaite versionner son projet depuis son poste de travail en ayant recours pour cela à un dépôt distant, disponible sur un espace type `Gitlab` ou `Github`. 

L'utilisation du logiciel `Git` en local, qui relève plutôt de l'usage quotidien du logiciel, est reportée dans une fiche dédiée ([Utiliser Git avec RStudio](#git)). 

::: recommandation
* Pour interagir avec un dépôt `Git` distant, l'utilisateur a le choix
entre deux protocoles d'authentification: `SSH` ou `HTTPS`.
* Il est conseillé d'utiliser le protocole  `HTTPS` qui est simple à configurer. 
Vous pouvez utiliser le protocole `SSH` si vous le souhaitez, mais il peut s'avérer difficile à configurer.
* Si vous utilisez  l'authentification `HTTPS`, il est recommandé 
de s'authentifier avec des jetons personnels d'accès (*personnel access tokens*) plutôt qu'avec 
un _login_ et un mot de passe.
:::
```

## Les différentes forges utilisables

> Une `r with_def("forges")` est un espace informatique en ligne qui accueille des dépôts distants. Un dépôt distant (_remote repository_ en anglais) est un dossier que vous pouvez utiliser pour stocker vos codes (et leur historique), et pour partager des codes avec vos collègues ou stocker vos codes personnels.

Reviewer 2: ```suggestion
Une `r with_def("forge")` est un espace informatique en ligne qui accueille des dépôts distants. Un dépôt distant (_remote repository_ en anglais) est un dossier que vous pouvez utiliser pour stocker vos codes (et leur historique), et pour partager des codes avec vos collègues ou stocker vos codes personnels.
```

L'authentification consiste ainsi à fournir un élément que seul vous et 
la forge sont censés connaître : un mot de passe, une clé compliquée, un jeton d'accès...

Plus précisément, il existe deux modalités pour faire connaître son identité à `Gitlab` :

* une **authentification HTTPS** ([décrite ici](git-connexion-https)) : l'authentification se fait grâce au login et au mot de passe, qu'il faut renseigner à chaque interaction avec le dépôt.
* **une authentification SSH** : l'authentification se fait par une clé cryptée disponible sur le poste de travail et que `Github` ou `Gitlab` connaît. Une fois configurée, cette méthode ne nécessite plus de faire connaître son identité : l'empreinte digitale que constitue la clé suffit à reconnaître un utilisateur. 

::: conseil
Si vous utilisez l'authentification HTTPS, il est conseillé d'utiliser un jeton d'accès. Celui-ci permet de s'authentifier, est facilement révocable et ne bénéficie pas des super-pouvoirs qu'octroie un mot de passe comme changer le nom d'un dépôt voire le supprimer. Plus d'éléments sur les jetons d'accès sont disponibles [ici](https://happygitwithr.com/credential-caching.html#get-a-pat).
:::

> ::: remarque
> Ne jamais stocker un token, et encore moins son mot de passe, dans un projet. Un token peut être stocké dans un système de gestion de mot de passe comme [Keepass](https://keepass.fr/)
> :::
> 

Reviewer 2: ```suggestion
::: remarque
Ne jamais stocker un _token_, et encore moins son mot de passe, dans un projet. Un _token_ peut être stocké dans un système de gestion de mot de passe comme [Keepass](https://keepass.fr/).
:::

```

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/clone_https.png", compression = FALSE)
```

Au moment du clônage, l'utilisateur est invité à entrer son identifiant
(celui de son profil Gitlab avec lequel il accède au dépôt distant), ainsi que
son mot de passe (celui qu'il aura donc spécifié dans son compte Gitlab).

::: conseil
Selon le niveau de visibilité fixé par le propriétaire du dépôt distant, il est possible qu'il ne soit pas nécessaire de s'identifier au moment du clônage (c'est le cas pour les dépôts publics). En revanche, si le niveau de visibilité du dépôt est plus élevé (niveau privé), il se peut qu'il soit impossible à l'utilisateur de clôner le dépôt ; dans ce cas, il ne lui sera pas non plus possible de visualiser ce dernier dans Gitlab. Il faut alors contacter le propriétaire du dépôt pour lui en demander l'accès.
:::

### Définir un jeton d'accès sur Gitlab

Dans `Gitlab`, en haut à droite de la page, cliquer sur `Edit profile` :

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/config_pat_gitlab1.png", compression = FALSE)
```

Dans le bandeau à gauche, cliquer sur `Access Tokens` :

> ```{r, echo = FALSE, out.width = '50%'}
> utilitr::include_image("./pics/config_git/config_pat_gitlab2.png", compression = FALSE)
> ```

Reviewer 2: ```suggestion
```{r, echo = FALSE, out.width = '20%'}
utilitr::include_image("./pics/config_git/config_pat_gitlab2.png", compression = FALSE)
```
```

utilitr::include_image("./pics/config_git/config_pat_gitlab3.png", compression = FALSE)
```

Après avoir validé, le jeton s'affiche afin de la récupérer et le stocker dans
un endroit sécurisé (par exemple un gestionnaire de mot de passes). Attention,
ce sera la seule fois qu'il sera visible.

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/config_pat_gitlab4.png", compression = FALSE)
```

Plus bas sur la page, il est possible de voir l'ensemble des jetons utilisés, 
la date de dernière utilisation. C'est aussi ici que peut être supprimé le 
jeton en cas de doute sur sa confidentialité. 

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/config_pat_gitlab5.png", compression = FALSE)
```

::: remarque
La marche à suivre pour créer un jeton d'accès sur `Github` est identique. La
[documentation officielle](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) présente des captures d'écran qui illustrent la démarche. Il faut choisir le 
niveau `repo` pour avoir les droits mentionnés précédemment. 
:::

### Garder en mémoire ses identifiants

> L'authentification `HTTPS` implique en principe de renseigner ses identifiants (login et
> password, ou token) à chaque interaction avec le dépôt distant. Toutefois, `Git` propose
> un outil pour conserver temporairement en mémoire (*cache*) des informations
> d'authentification. **Cette mise en mémoire est complètement optionnelle, et vous pouvez très bien utiliser  `Git` quotidiennement sans vous servir de cette possibilité.**

Reviewer 2: ```suggestion
L'authentification `HTTPS` implique en principe de renseigner ses identifiants (_login_ et
_password_, ou _token_) à chaque interaction avec le dépôt distant. Toutefois, `Git` propose
un outil pour conserver temporairement en mémoire (*cache*) des informations
d'authentification. **Cette mise en mémoire est complètement optionnelle, et vous pouvez utiliser  `Git` quotidiennement sans vous servir de cette possibilité.**
```

([explications ici](#terminal-git)), taper

~~~~shell
# windows
git config --global credential.helper wincred

# mac et linux
git config --global credential.helper 
~~~~

pour stocker pendant quinze minutes les informations de connexion.

En cas d'erreur
de frappe dans le mot de passe ou le jeton, il est possible de vider la mémoire
de la manière suivante:

~~~~shell
git config --global --unset credential.helper
~~~~

## Configurer l'accès à dépôt distant en `SSH` {#git-connexion-ssh}

Pour interagir avec un dépôt distant en utilisant le protocole `SSH`,
il faut créer une clé SSH, plus exactement une paire de chaînes de caractères,
qui va permettre une authentification automatique auprès du dépôt distant.
La version publique de la paire de clés est connue du dépôt distant ; la
version privée reste quant à elle la propriété seule et unique de
l'utilisateur. C'est l'association par un logiciel de cryptographie
de ces deux versions qui permet l'authentification de l'utilisateur.
::: specificite
L'authentification SSH ne fonctionne pas dans le Gitlab du SSP Cloud. Seule l'authentification en `HTTPS` est possible.
:::

::: specificite

> La documentation d'AUS propose un tutoriel détaillé sur La configuration de la clé SSH. Vous pouvez le trouver ici : `Y:/Documentation/AUSV3/Git_Utiliser Git sous AUSv3.pdf`.

Author: ```suggestion
La documentation d'AUS propose un tutoriel détaillé sur la configuration de la clé SSH. Vous pouvez le trouver ici : `Y:/Documentation/AUSV3/Git_Utiliser Git sous AUSv3.pdf`.
```

# Configurer `Git` sur son poste de travail {#git-config}

## Tâches concernées et recommandations

L'utilisateur souhaite versionner son projet depuis son poste de travail en ayant recours pour cela à un dépôt distant, disponible sur un espace type `Gitlab` ou `Github`. 

L'utilisation du logiciel `Git` en local, qui relève plutôt de l'usage habituel du logiciel, est reportée dans une fiche dédiée ([Utiliser Git avec RStudio](#git)). 

::: recommandation
* Pour interagir avec un dépôt `Git` distant, l'utilisateur a le choix
entre deux protocoles de connexion, `SSH` ou `HTTPS`: 
* Le
protocole `SSH` est déconseillé en faveur du protocole `HTTPS` (qui ne requiert
que de configurer un mot de passe sur le dépôt distant) ;
* Plutôt qu'utiliser un mot de passe pour s'authentifier en `HTTPS`, il est
recommandé de s'authentifier avec des jetons personnels d'accès
(*personnel access tokens*).
:::

> 

Author: Alors, je retire, on a bien la possibilité de faire du HTTPS (et donc d'utiliser des jetons) avec le gitlab interne et AUS. Je viens de tester, ça fonctionne maintenant. 

Le HTTPS est la méthode la plus universelle (ça marche partout, vers n'importe quelle plateforme contrairement au SSH qui est bloqué vers gitlab.com ou github.com). C'est la méthode recommandée par Github. C'est sécurisé et ça be demande pas de configuration, ce qui est assez attractif. Et l'aspect le plus pénible, à savoir rentrer son mot de passe à chaque fois, peut être évité grâce au stockage des *credentials* dans la mémoire *cache*. Je crois que @RLesur est du même avis, il aura peut-être des arguments supplémentaires

> 

Author: Après honnêtement, je suis pas un anti SSH. Je l'utilise aussi. Mais je trouve que c'est plus difficile à comprendre conceptuellement (clés privées et publiques) que le HTTPS,  plus lourd à configurer et moins universel

version publique mais proposent des machines (appelées *runners*) internes. Celles-ci ne sont pas forcément ouvertes sur internet.   

Ces dépôts distants, également appelés `r with_def("forges")`, proposent des fonctionalités précieuses pour la gestion d'un projet impliquant du code ou de la documentation. La formation
[Travail collaboratif avec R](https://linogaliana.gitlab.io/collaboratif/git.html) évoque de manière plus extensive les fonctionalités.

::: specificite

Une instance interne `Gitlab` est disponible sous AUS. Son adresse, ainsi que des éléments complémentaires à cette fiche, sont disponibles dans la documentation AUS (`Y:/Documentation/AUSV3/`)

:::

::: remarque

Une instance `Gitlab` est disponible sur le SSP-cloud [https://git.lab.sspcloud.fr/] avec laquelle peut interagir un service `RStudio` [lien vers fiche quand elle sera là] . Cependant, cette plateforme étant connectée à Internet, elle a également accès aux dépôts sur  <https://gitlab.com> ou <https://github.com> dont la pérennité est mieux assurée. Il est ainsi conseillé de privilégier ces deux `r with_def("forges")` plutôt que <https://git.lab.sspcloud.fr/>

> :::

Author: T'es sûr de mettre ça dans une box spécificité ? On est pas vraiment sur une spécificité Insee là

> :::

Reviewer 2: OK, on peut en faire une remarque.

* une authentification SSH: l'authentification se fait par une clé cryptée disponible sur le poste de travail et que `Github` ou `Gitlab` connaît. Une fois configurée, cette méthode ne nécessite plus de faire connaître son identité: l'empreinte digitale que constitue la clé suffit à reconnaître un utilisateur. 

::: remarque
Ne jamais stocker un token, et encore moins son mot de passe, dans un projet. Un token peut être stocké dans un système de gestion de mot de passe comme [Keepass](https://keepass.fr/)
:::

La méthode SSH peut être laborieuse à mettre en place même si elle donne un 
confort supplémentaire (ne pas avoir à taper login et mot de passe à chaque interaction). Cependant, la méthode HTTPS est préférable car

* elle n'implique pas de configuration ;
* elle n'est pas bloquée par des pare-feux, contrairement à la méthode SSH. Par exemple, il est impossible sur un poste Insee d'interagir avec un dépôt sur <https://gitlab.com> ou <https://github.com> en SSH ;
* il est possible de conserver en mémoire ses identifiants un certain temps (grâce au `git credentials helper`) pour ne pas taper à chaque fois ses identifiants.

> ## Clôner et interagir avec un dépôt distant en `HTTPS` {#git-connexion-https}
> 

Reviewer 2: ```suggestion
## Configurer l'accès à dépôt distant en `HTTPS` {#git-connexion-https}

```

> ## Clôner et interagir avec un dépôt distant en `HTTPS` {#git-connexion-https}
> 
> ::: specificite
> 
> Le `Gitlab` interne de l'Insee ne permet pas ce mode d'authentification. 
> Il convient d'utiliser l'[authentification SSH](#git-connexion-ssh)
> 
> :::

Reviewer 2: ```suggestion
## Clôner et interagir avec un dépôt distant en `HTTPS` {#git-connexion-https}

::: specificite
Le `Gitlab` interne de l'Insee ne permet pas ce mode d'authentification. 
Il convient d'utiliser l'[authentification SSH](#git-connexion-ssh).
:::
```

qu'à `r with_def("cloner")` le dépôt distant sur lequel on veut travailler,
par exemple depuis `RStudio`, suivant le protocole décrit dans la fiche [Utiliser Git avec RStudio](#git), en ayant pris soin de sélectionner l'URL correspondant
au protocole `HTTPS`, comme ci-dessous :

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/clone_https.png", compression = FALSE)
```

Au moment du clônage, l'utilisateur est invité à entrer son identifiant
(celui de son profil Gitlab avec lequel il accède au dépôt distant), ainsi que
son mot de passe (celui qu'il aura donc spécifié dans son compte Gitlab).

::: conseil
Selon le niveau de visibilité fixé par le propriétaire du dépôt distant, il est possible qu'il ne soit pas nécessaire de s'identifier au moment du clônage (c'est le cas pour les dépôts publics). En revanche, si le niveau de visibilité du dépôt est plus élevé (niveau privé), il se peut qu'il soit impossible à l'utilisateur de clôner le dépôt ; dans ce cas, il ne lui sera pas non plus possible de visualiser ce dernier dans Gitlab. Il faut alors contacter le propriétaire du dépôt pour lui en demander l'accès.
:::

> ### Définir un jeton d'accès
> 
> #### Sur Gitlab
> 

Reviewer 2: ```suggestion
### Définir un jeton d'accès sur Gitlab

```

> #### Sur Gitlab
> 
> Dans `Gitlab`, en haut à droite de la page, cliquer sur `Edit profile`
> 

Reviewer 2: ```suggestion
#### Sur Gitlab

Dans `Gitlab`, en haut à droite de la page, cliquer sur `Edit profile` :

```

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/config_pat_gitlab1.png", compression = FALSE)
```

> Dans le bandeau à gauche, cliquer sur `Access Tokens`:
> 

Reviewer 2: ```suggestion
Dans le bandeau à gauche, cliquer sur `Access Tokens` :

```

```{r, echo = FALSE, out.width = '50%'}
utilitr::include_image("./pics/config_git/config_pat_gitlab2.png", compression = FALSE)
```

> Choisir un nom (par exemple `PAT_GITLAB`, mais ce n'est pas un nom obligatoire) et
> des droits associés. Les droits `read_repository` et `write_repository`
> permettent d'effectuer les routines décrites dans la fiche 
> [Utilisation Git avec RStudio](#git).
> 

Reviewer 2: ```suggestion
Choisir un nom (par exemple `PAT_GITLAB`, mais ce n'est pas un nom obligatoire) et
des droits associés. Les droits `read_repository` et `write_repository`
permettent d'effectuer les opérations standard (_pull_ et _push_) décrites dans la fiche 
[Utilisation Git avec RStudio](#git).

```

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/config_pat_gitlab3.png", compression = FALSE)
```

Après avoir validé, le jeton s'affiche afin de la récupérer et le stocker dans
un endroit sécurisé (par exemple un gestionnaire de mot de passes). Attention,
ce sera la seule fois qu'il sera visible.

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/config_pat_gitlab4.png", compression = FALSE)
```

Plus bas sur la page, il est possible de voir l'ensemble des jetons utilisés, 
la date de dernière utilisation. C'est aussi ici que peut être supprimé le 
jeton en cas de doute sur sa confidentialité. 

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/config_pat_gitlab5.png", compression = FALSE)
```

> 
> Sur `Github`, la démarche est identique. La
> [documentation officielle](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token)
> présente les captures d'écran pour illustrer la démarche. Il faut choisir le 
> niveau `repo` pour avoir les droits mentionnés précédemment. 
> 
> 

Reviewer 2: ```suggestion

::: remarque
La marche à suivre pour créer un jeton d'accès sur `Github` est identique. La
[documentation officielle](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) présente des captures d'écran qui illustrent la démarche. Il faut choisir le 
niveau `repo` pour avoir les droits mentionnés précédemment. 
:::

```

> ### Garder en mémoire (*cache*) ses identifiants
> 
> Le système `HTTPS` implique, en principe, de rappeler login et
> password (ou token) à chaque authentification. Cependant, `Git` propose
> un outil pour mettre en mémoire temporaire (*cache*) des informations
> d'authentification. Il convient de passer par la ligne de commande. 
> 
> Après avoir ouvert une invite de commande `Git Bash`
> ([explications ici](#terminal-git)), taper
> 
> ~~~~shell
> # windows
> git config --global credential.helper wincred
> 
> # mac et linux
> git config --global credential.helper 
> ~~~~
> 
> 
> pour stocker pendant quinze minutes les informations de connexion.
> 
> En cas d'erreur
> de frappe dans le mot de passe ou le jeton, il est possible de vider la mémoire
> de la manière suivante:
> 
> ~~~~shell
> git config --global --unset credential.helper
> ~~~~

Reviewer 2: ```suggestion
### Garder en mémoire ses identifiants

L'authentification `HTTPS` implique en principe de renseigner ses identifiants (login et
password, ou token) à chaque interaction avec le dépôt distant. Toutefois, `Git` propose
un outil pour conserver temporairement en mémoire (*cache*) des informations
d'authentification. **Cette mise en mémoire est complètement optionnelle, et vous pouvez très bien utiliser  `Git` quotidiennement sans vous servir de cette possibilité.**

La mise en mémoire se fait en deux temps:

- Il faut d'abord ouvrir une invite de commande `Git Bash`
([explications ici](#terminal-git)) ;
- Il faut ensuite exécuter la ligne de code suivante pour stocker pendant quinze minutes les informations de connexion :

    ~~~~shell
    # Sous windows
    git config --global credential.helper wincred
    
    # Sous mac et linux
    git config --global credential.helper 
    ~~~~

::: remarque
Si vous avez fait une faute de frappe dans le mot de passe ou dans le jeton, il est possible de vider la mémoire
de la manière suivante :

~~~~shell
git config --global --unset credential.helper
~~~~
:::
```

> 
> ## Clôner et interagir avec un dépôt distant en `SSH` {#git-connexion-ssh}
> 

Reviewer 2: ```suggestion

## Configurer l'accès à dépôt distant en `SSH` {#git-connexion-ssh}

```

Pour interagir avec un dépôt distant en utilisant le protocole `SSH`,
il faut créer une clé SSH, plus exactement une paire de chaînes de caractères,
qui va permettre une authentification automatique auprès du dépôt distant.
La version publique de la paire de clés est connue du dépôt distant ; la
version privée reste quant à elle la propriété seule et unique de
l'utilisateur. C'est l'association par un logiciel de cryptographie

> de ces deux versions qui permet l'authentification de l'utilisateur.

Reviewer 2: Je pense que ce serait bien de faire une petite section ### sur "c'est quoi SSH, et la distinction entre clé publique et clé privée", puis dire qu'il y a deux étapes : créer la clé SSH, et l'enregistrer dans la forge.

> 

Reviewer 2: ```suggestion
::: specificite
L'authentification SSH ne fonctionne pas dans le Gitlab du SSP Cloud. Seule l'authentification en `HTTPS` est possible.
:::

```

> ### Générer une clé SSH
> 
> Un utilisateur peut générer autant de clés SSH que bon lui semble, selon qu'il utilise plusieurs systèmes par exemple.
> 
> Pour générer une clé SSH, le plus simple est d'utiliser `Git Bash`. Après avoir
> ouvert une invite de commande, taper :
> 

Reviewer 2: ```suggestion
### Générer une clé SSH

Un utilisateur peut générer autant de clés SSH qu'il le souhaite. Par exemple, vous pouvez générer une clé SSH par projet sur lequel vous travaillez (même si l'utilité de cette méthode n'est pas évidente).

Voici comment créer une clé SSH avec `Git Bash` :

- Il faut d'abord ouvrir une invite de commande `Git Bash`
([explications ici](#terminal-git)) ;
- Il faut ensuite exécuter la ligne de code suivante :
```

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/ssh_keygen.png", compression = FALSE)
```

> La clé est générée sur le chemin spécifié et sera constituée de deux fichiers texte :
> - un fichier en `.pub`, qui fournit la partie publique de la clé, et qui a vocation à être diffusée (donc c'est ce fichier qu'on chargera sur Gitlab pour permettre l'authentification) ;
> - un fichier `.ppk` (comme Private Key) qui est la partie privée de la clé, et qui ne doit en aucun cas être diffusée.
> 
> La version privée de la clé doit être localisée dans un dossier caché que
> Git va automatiquement utiliser
> (en général un dossier `.ssh` localisé sur un dossier personnel).

Reviewer 2: ```suggestion
La clé est générée sur le chemin spécifié et sera constituée de deux fichiers texte :

- un fichier dont le nom se termine par `.pub`, qui fournit la partie publique de la clé, et qui a vocation à être diffusée (donc c'est ce fichier qu'on chargera sur Gitlab pour permettre l'authentification) ;
- un fichier dont le nom se termine par `.ppk` (comme Private Key) qui est la partie privée de la clé, et qui ne doit en aucun cas être diffusée.

La version privée de la clé doit être localisée dans un dossier caché que Git va automatiquement utiliser (en général un dossier `.ssh` localisé sur un dossier personnel).

```

### Ajout de la clé SSH sous Gitlab

Pour copier la partie publique de la clé sous Gitlab, il faut aller dans
la partie `Settings` de l'espace personnel sous Gitlab, puis sélectionner
*"SSH Keys"* sur le bandeau latéral gauche. On se trouve face à une fenêtre
de texte comme ceci :

```{r, echo = FALSE, out.width = '100%'}
utilitr::include_image("./pics/config_git/ssh_key_record.png", compression = FALSE)
```

dans laquelle on doit copier-coller le texte que l'on trouve dans le
fichier `.pub` de la clé SSH (en prenant soin d'éliminer l'éventuel texte
en fin du type *"imported-openssh-key"*).

Gitlab est désormais configuré, et il est alors possible de faire
dialoguer Git et Gitlab en suivant, par exemple, les routines proposées dans
la fiche [Utiliser R avec RStudio](#git).

> ::: conseil
> L'option SSH ne fonctionne pas dans le Gitlab du SSP Cloud.
> Seule l'authentification en `HTTPS` est possible.
> :::
> 

Reviewer 2: ```suggestion

```

::: conseil
Si vous utilisez l'authentification HTTPS, il est conseillé d'utiliser un jeton d'accès. Celui-ci permet de s'authentifier, est facilement révocable et ne bénéficie pas des super-pouvoirs qu'octroie un mot de passe comme changer le nom d'un dépôt voire le supprimer. Plus d'éléments sur les jetons d'accès sont disponibles [ici](https://happygitwithr.com/credential-caching.html#get-a-pat).
:::

::: remarque
Ne jamais stocker un _token_, et encore moins son mot de passe, dans un projet.
Il est possible de stocker un mot de passe ou *token* de manière sécurisée et durable
avec le *credential helper* de `Git`. Celui-ci est présenté par la suite.

S'il n'est pas possible d'utiliser le *credential helper* de `Git`, un mot de passe 
ou _token_ peut être stocké de manière sécurisé dans
un système de gestion de mot de passe comme [Keepass](https://keepass.fr/).
:::

> La méthode SSH peut être laborieuse à mettre en place, mais elle est plus confortable dans le travail quotidien car elle permet de ne pas avoir à renseigner son login et son mot de passe à chaque interaction avec le dépôt distant. Cependant, la méthode HTTPS est préférable car :

Author: ```suggestion
La méthode SSH peut être laborieuse à mettre en place, mais elle confortable dans le travail quotidien car elle permet de ne pas avoir à renseigner son login et son mot de passe à chaque interaction avec le dépôt distant. Cependant, la méthode HTTPS est préférable car :
```

niveau `repo` pour avoir les droits mentionnés précédemment. 
:::

### Garder en mémoire ses identifiants

L'authentification `HTTPS` implique en principe de renseigner ses identifiants (_login_ et
_password_, ou _token_) à chaque interaction avec le dépôt distant. Toutefois, `Git` propose
un outil pour conserver temporairement en mémoire (*cache*) des informations
d'authentification. **Cette mise en mémoire est complètement optionnelle, et vous pouvez utiliser  `Git` quotidiennement sans vous servir de cette possibilité.**

La mise en mémoire se fait en deux temps:

- Il faut d'abord ouvrir une invite de commande `Git Bash`
([explications ici](#terminal-git)) ;
- Il faut ensuite exécuter la ligne de code suivante pour stocker pendant quinze minutes les informations de connexion :

    ~~~~shell
    # Sous windows
    git config --global credential.helper manager-core
    
    # Sous mac et linux
    git config --global credential.helper 
    ~~~~

::: remarque
Si vous avez fait une faute de frappe dans le mot de passe ou dans le jeton, il est possible de vider la mémoire

> de la manière suivante :

Author: ```suggestion
de la manière suivante, sous Mac ou Linux :
```

~~~~shell
git config --global --unset credential.helper

> ~~~~

Author: ```suggestion
~~~~

Sous Windows, si vous avez utilisé l'option `manager-core` évoquée ci-dessus, vous pouvez utiliser une interface graphique pour effacer le mot de passe ou jeton erroné. Pour cela, dans le menu démarrer, taper `Gestionnaire d'identification` (ou `Credential Manager` si Windows ne trouve pas). Dans l'interface graphique qui s'ouvre, il est possible de supprimer le mot de passe ou jeton en question. Après cela, vous devriez à nouveau avoir l'opportunité de taper un mot de passe ou jeton lors d'une authentification HTTPS.
```
